Frameworks/Foundation/NSUndoManager.mm

```diff
@@ -38,83 +104,247 @@ - (id)prepareWithInvocationTarget:(id)target {
+    }
+    if (!_redoing) {
+        _groupingLevel = _undoGroupingLevel;
+    }
 }
 
 /**
- @Status Stub
- @Notes
+ @Status Interoperable
 */
 - (void)endUndoGrouping {
-    UNIMPLEMENTED();
+    if (_undoing) {
+        return;
+    }
+    // No group to close.
+    if (_undoGroupingLevel == 0) {
+        [NSException raise:NSInternalInconsistencyException format:@"endUndoGrouping called with no open group."];
+    }
+    _NSUndoGroup* undoGroup = [_undoStack peek];
+    [undoGroup closeUndoGroup];
+    _undoGroupingLevel--;
+    if (!_redoing) {
+        _groupingLevel = _undoGroupingLevel;
+    }
+}
+
+// Internal redo grouping
+- (void)_beginRedoGrouping {
+    _NSUndoGroup* redoGroup = [_redoStack peek];
+
+    if (redoGroup == nil || [redoGroup isClosed]) {
+        _NSUndoGroup* newGroup = [[_NSUndoGroup alloc] init];
+        [_redoStack push:newGroup];
+        [newGroup release];
+    } else {
+        [redoGroup createUndoGroup];
+    }
+    _redoGroupingLevel++;
+    if (!_redoing) {
+        _groupingLevel = _redoGroupingLevel;
+    }
+}
+
+- (void)_endRedoGrouping {
+    // No group to close.
+    if (_redoGroupingLevel == 0) {
+        [NSException raise:NSInternalInconsistencyException format:@"Attempted to close a redo group with no open redo group."];
+    }
+    _NSUndoGroup* redoGroup = [_redoStack peek];
+    [redoGroup closeUndoGroup];
+    _redoGroupingLevel--;
+    if (!_redoing) {
+        _groupingLevel = _redoGroupingLevel;
+    }
 }
 
 /**
- @Status Stub
- @Notes
+@Status Interoperable
+*/
+- (BOOL)isUndoRegistrationEnabled {
+    @synchronized(self) {
+        return _disableCount == 0;
+    }
+}
+
+/**
+ @Status Interoperable
 */
 - (void)disableUndoRegistration {
-    UNIMPLEMENTED();
+    @synchronized(self) {
+        _disableCount++;
+    }
 }
 
 /**
- @Status Stub
- @Notes
+ @Status Interoperable
 */
 - (void)enableUndoRegistration {
-    UNIMPLEMENTED();
+    @synchronized(self) {
+        if (_disableCount == 0) {
+            [NSException raise:NSInternalInconsistencyException
+                        format:@"Called enable undo registration while undo registration is enabled."];
+        }
+        _disableCount--;
+    }
 }
 
 /**
- @Status Stub
- @Notes
+ @Status Interoperable
 */
 - (void)removeAllActions {
-    UNIMPLEMENTED();
+    [self _removeUndoActions];
+    [self _removeRedoActions];
+
+    _undoGroupingLevel = 0;
+    _redoGroupingLevel = 0;
+    _isAutomaticGroupOpen = NO;
+    _hasNestedGroup = NO;
+}
+
+- (void)_removeUndoActions {
+    _undoStack.attach([[_NSUndoManagerStack alloc] init]);
+}
+
+- (void)_removeRedoActions {
+    _redoStack.attach([[_NSUndoManagerStack alloc] init]);
 }
 
 /**
- @Status Stub
+ @Status Interoperable
  @Notes
 */
 - (void)removeAllActionsWithTarget:(id)target {
-    UNIMPLEMENTED();
+    [_undoStack removeAllWithTarget:target];
+    [_redoStack removeAllWithTarget:target];
+
+    _undoGroupingLevel = [[_undoStack peek] updateLevel];
```

Comment:
>updateLevel [](start = 44, length = 11)

Is this called `updateLevel` because it is a level of updates, or because it will update the level.

If the first, why are we talking about updates? If the second, it doesn't look like it's setting anything... it isn't performing an update.